### PR TITLE
Notify push and email on code exchanges

### PR DESCRIPTION
**Pull request by vladikoff**
Fixes https://github.com/mozilla/fxa-auth-server/issues/2880
Fixes https://github.com/mozilla/fxa-auth-server/issues/2955

### CODE REVIEW

File: fxa-oauth-server/lib/routes/token.js [hunk 1]
```diff
@@ -162,7 +162,7 @@ module.exports = {
       access_token: validators.token.required(),
       refresh_token: validators.token,
       id_token: validators.assertion,
-      scope: validators.scope.required(),
+      scope: validators.scope.required().allow(''), // YA!
```
Comment (rfk) [inline fxa-oauth-server/lib/routes/token.js]: I'm going to try changing the scope regex to use `*` rather than `+`, so we accept the empty string everywhere we validate a scope.

File: lib/devices.js [hunk 2]
```diff
@@ -114,7 +114,7 @@ module.exports = (log, db, push) => {
             deviceName = synthesizeName(deviceInfo)
           }
           if (credentials.tokenVerified) {
-            request.app.devices.then(devices => {
+            db.devices(credentials.uid).then(devices => {
```
Comment (rfk) [inline lib/devices.js]: Why was this needed? I wonder if we can find a way to fix `request.app.devices` so that we maintain the nice caching behaviour.

File: lib/routes/oauth.js [hunk 3]
```diff
@@ -55,6 +58,82 @@ module.exports = (log, config, oauthdb) => {
...
+            grant = await oauthdb.grantTokensFromSessionToken(sessionToken, request.payload)
+            break
+          default:
+            throw error.internalValidationError()
+        }
+
+        switch (request.payload.grant_type) {
+          case 'authorization_code':
+          case 'fxa-assertion':
+            await oauthRouteUtils.newTokenNotification(db, oauthdb, mailer, devices, request, sessionToken, grant)
```
Comment (rfk) [inline lib/routes/oauth.js]: Rather than switching on `grant_type`, I wonder if what you want here is to switch on whether a `refresh_token` was generated.

File: lib/routes/utils/oauth.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    if (! scopeSet.intersects(NOTIFICATION_SCOPES)) {
+      // right now we only care about notifications for the `oldsync` scope
+      // if not a match, then we don't do any notifications
+      return
+    }
+
+    const tokenVerify = await oauthdb.checkAccessToken({
+      token: grant.access_token
+    })
```
Comment (rfk) [inline lib/routes/utils/oauth.js]: It seems like a bit of a code smell that you have to check the access_token here immediately after generating it; is this so that you can find out the `uid` in the `grant_type=authorization_code` case?  I wonder if we could arrange for the oauth-server to return that information directly from `/token` rather than having to verify it again here.

Comment (vladikoff) [reply lib/routes/utils/oauth.js]: Yeah that would be nice :)

File: fxa-oauth-server/config/dev.json [hunk 5]
```diff
@@ -144,7 +144,7 @@
       "hashedSecret": "a7ee3482fab1782f5d3945cde06bb911605a8dfc1a45e4b77bc76615d5671e51",
       "imageUri": "",
       "redirectUri": "http://127.0.0.1:3030/oauth/success/3c49430b43dfba77",
-      "canGrant": false,
+      "canGrant": true,
```
Comment (vladikoff) [inline fxa-oauth-server/config/dev.json]: Fenix and Reference browser need this in prod and elsewhere when this lands. fxa-dev boxes

File: fxa-oauth-server/lib/routes/token.js (hunk 1 above)
Comment (vladikoff) [reply fxa-oauth-server/lib/routes/token.js]: this change should go away, fix in the other PR to the regex.

File: lib/oauthdb/check-access-token.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const Joi = require('joi')
+const validators = require('../routes/validators')
+
+module.exports = (config) => {
+  return {
+    path: '/v1/verify',
```
Comment (vladikoff) [inline lib/oauthdb/check-access-token.js]: get rid of this, `/token` endpoint can provide the uid

Comment (vladikoff) [reply lib/oauthdb/check-access-token.js]: `fxa-uid` field

File: lib/oauthdb/grant-tokens-from-assertion.js [hunk 7]
```diff
@@ -26,8 +26,9 @@ module.exports = (config) => {
         access_token: validators.accessToken.required(),
         refresh_token: validators.refreshToken.optional(),
         id_token: validators.assertion.optional(),
-        scope: validators.scope.required(),
+        scope: validators.scope.required().allow(''), // TODO: needs the ''?, had this problem before
         token_type: Joi.string().valid('bearer').required(),
+        auth_at: Joi.number().optional(), // TODO: required?
```
Comment (vladikoff) [inline lib/oauthdb/grant-tokens-from-assertion.js]: fixed in https://github.com/mozilla/fxa-auth-server/pull/2983/files#r269313254

File: lib/routes/oauth.js (hunk 3 above)
Comment (vladikoff) [reply lib/routes/oauth.js]: `grant.refreshToken`

File: lib/routes/utils/oauth.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    if (! credentials) {
+      credentials = {}
+    }
+
+    const uid = tokenVerify.user
+    credentials.uid = uid
+    credentials.tokenVerified = true // XXX TODO: check this?
```
Comment (vladikoff) [inline lib/routes/utils/oauth.js]: need a comment for oldsync scope

File: lib/routes/utils/oauth.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    credentials.refreshTokenId = grant.refresh_token
```
Comment (vladikoff) [inline lib/routes/utils/oauth.js]: check for null here i guess?

Comment (vladikoff) [reply lib/routes/utils/oauth.js]: object assign?

File: lib/routes/utils/oauth.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    credentials.client = await oauthdb.getClientInfo(tokenVerify.client_id)
+
+    await devices.upsert(request, credentials, {})
```
Comment (vladikoff) [inline lib/routes/utils/oauth.js]: no names, but client name from `credentials` provides a name for the device

File: lib/routes/utils/oauth.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+    const geoData = request.app.geo
+    const ip = request.app.clientAddress
+    const service = request.payload.service || request.query.service || tokenVerify.client_id
```
Comment (vladikoff) [inline lib/routes/utils/oauth.js]: should be in `request`., tests might have been wrong!

Comment (vladikoff) [reply lib/routes/utils/oauth.js]: just do `client_id` here

File: lib/routes/utils/oauth.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+      acceptLanguage: request.app.acceptLanguage,
+      ip: ip,
+      location: geoData.location,
+      service: service,
+      timeZone: geoData.timeZone,
+      uid: credentials.uid
+    }
+
+    const account = await db.account(uid)
+    await mailer.sendNewDeviceLoginNotification(account.emails, account, emailOptions)
```
Comment (vladikoff) [inline lib/routes/utils/oauth.js]: might have emails on sessionToken

File: lib/devices.js (hunk 2 above)
Comment (vladikoff) [reply lib/devices.js]: it's problematic, https://github.com/mozilla/fxa-auth-server/issues/2992 was filed.

File: lib/routes/utils/oauth.js (hunk 4 above)
Comment (vladikoff) [reply lib/routes/utils/oauth.js]: @rfk how about we leave this as a follow up for now? merge with the `checkAccessToken` then optimize by landing https://github.com/mozilla/fxa-auth-server/pull/3000 ?

File: lib/oauthdb/check-access-token.js (hunk 6 above)
Comment (vladikoff) [reply lib/oauthdb/check-access-token.js]: in https://github.com/mozilla/fxa-auth-server/pull/3000

File: lib/routes/utils/oauth.js (hunk 12 above)
Comment (vladikoff) [reply lib/routes/utils/oauth.js]: nope, no emails on sessionToken, just one email.

File: lib/routes/utils/oauth.js (hunk 4 above)
Comment (rfk) [reply lib/routes/utils/oauth.js]: OK, works for me.

File: lib/routes/utils/oauth.js [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+      // provide a refreshToken for the device creation below
+      credentials.refreshTokenId = grant.refresh_token;
+    }
+
+    // we set tokenVerified because the granted scope is part of NOTIFICATION_SCOPES
+    credentials.tokenVerified = true;
+    credentials.client = await oauthdb.getClientInfo(clientId);
+
+    // The following upsert gets no `deviceInfo`.
+    // However, `credentials` provides a `client` which generates the name for the device
```
Comment (rfk) [inline lib/routes/utils/oauth.js]: nit: it took me a little bit to understand what this comment was trying to say; it might be clearer worded along the lines of "However, `credentials.client` lets it generate a default name for the device".

File: test/client/api.js [hunk 14]
```diff
@@ -1021,6 +1021,18 @@ module.exports = config => {
     );
   };
 
+  ClientApi.prototype.grantTokensFromSessionToken = function (sessionTokenHex, oauthParams) {
```
Comment (rfk) [inline test/client/api.js]: This is already in `master` under a slightly different name, `grantOAuthTokensFromSessionToken`.

File: lib/routes/utils/oauth.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    if (! credentials.refreshTokenId) {
+      // provide a refreshToken for the device creation below
+      credentials.refreshTokenId = grant.refresh_token;
+    }
+
+    // we set tokenVerified because the granted scope is part of NOTIFICATION_SCOPES
+    credentials.tokenVerified = true;
+    credentials.client = await oauthdb.getClientInfo(clientId);
```
[the remark this answers is not in the document]
Comment (vladikoff) [reply lib/routes/utils/oauth.js]: 👍

File: lib/routes/utils/oauth.js [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+      const tokenVerify = await oauthdb.checkAccessToken({
+        token: grant.access_token
+      });
+      // some grant flows won't have the uid in `credentials`
+      credentials.uid = tokenVerify.user;
+    }
+
+    if (! credentials.refreshTokenId) {
+      // provide a refreshToken for the device creation below
+      credentials.refreshTokenId = grant.refresh_token;
```
[the remark this answers is not in the document]
Comment (vladikoff) [reply lib/routes/utils/oauth.js]: oh classic ... , updated , also updated the test

File: lib/routes/utils/oauth.js [hunk 17]
```diff
@@ -0,0 +1,62 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict';
+
+const encrypt = require('../../../fxa-oauth-server/lib/encrypt');
```
Comment (rfk) [inline lib/routes/utils/oauth.js]: Is this going to be OK at runtime in the docker images we build for production?  I wonder if it would be less risky to just copy the code over here, it's only 3 lines worth.

Comment (vladikoff) [reply lib/routes/utils/oauth.js]: Yeah! we have the technology!
